### PR TITLE
[FEATURE ember-contextual-components] Contextual components per RFC #64

### PR DESCRIPTION
**Pull request by Serabe**
emberjs/rfcs#64 ([rendered html](https://github.com/emberjs/rfcs/blob/master/text/0064-contextual-component-lookup.md))

Introduce new syntax and semantics for improving component composition.
## TODO
- [x] Extract current `component` helper implementation to `element_component`.
- [x] Add `ember-contextual-components` feature flag.
- [x] Implement `hash` helper.
- [x] Implement `closure_component`.
- [x] Make `component` helper handle a component closure.
- [x] Refactor positionalParams to be expression specific ~~(after #12350 this is no longer needed)~~
- [x] ~~Implement `{{my.component.closure}}` syntax.~~ This needs changes in HTMLBars.
- [x] Clean up the code a bit
- [x] ~~API docs for `(component)`~~ (left for a different PR after this has been merged)
- [x] Squash everything into one commit with the proper tag.

### CODE REVIEW

**Comment by rwjblue** [comment]
Lookin good so far, keep it up!

**Comment by Serabe** [comment]
After a great talk with @mixonic, it seems the best road is to do a component cell. Since my streams and cells knowledge is pretty limited, I tried to follow the `mut` example. In this [first draft of `closure-component`](https://github.com/Serabe/ember.js/blob/7bec093768147c947373b91f92ee0f0cf27a62d6/packages/ember-htmlbars/lib/keywords/closure-component.js) I try to do so but only as far as I think I understand what is happening in there.

I create a `ComponentStream` that looks quite similar to `MutStream`:

1. In the init, I keep the `componentPath`, `params` and `hash` referenced, as any of them can be streams or contain one. I set a proper `label` and add dependencies as well.

2. The `cell` method does what I think the `mut` version does: `read`ing (as in `streams/utils` `read`) the stream content.

3. I did not add an `[INVOKE]` method because I am not sure of what it does.

I did not plumbing yet because I am only going as far as I feel I understand properly.

File: packages/ember-htmlbars/lib/helpers/hash.js [hunk 1]
```diff
@@ -0,0 +1,27 @@
...
+@submodule ember-templates
+*/
+
+/**
+   Use the `{{hash}}` helper to create a hash to pass as an option to your
+   components. This is specially useful for contextual components where you can
+   just yield a hash:
+
+   ```handlebars
+   {{yield (hash
```
Comment (mixonic) [inline packages/ember-htmlbars/lib/helpers/hash.js]: We can give a more generic example here.

File: packages/ember-htmlbars/tests/helpers/hash_test.js [hunk 2]
```diff
@@ -0,0 +1,70 @@
+import EmberView from 'ember-views/views/view';
+import run from 'ember-metal/run_loop';
+import { set } from 'ember-metal/property_set';
+import compile from 'ember-template-compiler/system/compile';
+import { runAppend, runDestroy } from 'ember-runtime/tests/utils';
+
+var view;
+
+QUnit.module('hash helper', {
```
Comment (mixonic) [inline packages/ember-htmlbars/tests/helpers/hash_test.js]: Should be feature flagged

File: packages/ember-htmlbars/lib/main.js [hunk 3]
```diff
@@ -147,6 +148,7 @@ registerHelper('-normalize-class', normalizeClassHelper);
 registerHelper('concat', concatHelper);
 registerHelper('-join-classes', joinClassesHelper);
 registerHelper('-html-safe', htmlSafeHelper);
+registerHelper('hash', hashHelper);
```
Comment (mixonic) [inline packages/ember-htmlbars/lib/main.js]: Should be feature flagged (just the one flag for contextual components is good enough IMO)

**Comment by Serabe** [comment]
What is [this error](https://travis-ci.org/emberjs/ember.js/jobs/80931770) about?

In this [other error](https://travis-ci.org/emberjs/ember.js/jobs/80931770), I see a lot of them but then the numbers say everything is fine.

**Comment by mixonic** [comment]
@Serabe the first error is sauce labs. Either there is something here incompatible with IE9, or (more likely) sauce labs had an intermittent failure. I'll kick the server.

The "other" error, is the same link in your comment.

File: packages/ember-htmlbars/lib/helpers/hash.js (hunk 1 above)
Comment (mixonic) [inline packages/ember-htmlbars/lib/helpers/hash.js]: donezo

**Comment by mixonic** [comment]
@wycats @mmun @rwjblue I'd like to merge this PR with an initial implementation of contextual components. It does not include:

* Explicit GlimmerComponent support/tests, though I expect it is free.
* Support for `{{dot.path}}` being considered as a component. That will need HTMLBars effort.
* Packaging of dependencies. By this I mean the stream for the component cell should not invalidate when a hash/param value changes. Instead that should should be propagated through directly to the invocations. The invocation should be dependent upon the streams of the hash, we should not need to be dirtying the component cell stream. This is an optimization though, and we can fix it while people are experimenting with the core feature behind a flag.

Getting this behind a FF will allow devs to explore the feature and get us early feedback. I hope to tie up the loose ends here as glimmer component lands and ship with 2.3 (ideally).

@Serabe has also edited the component helper documentation to explain nested usage in a branch, so there are docs in the works already.

**Comment by mmun** [comment]
If it's feature flagged and works reasonably well I am :+1: merging.

**Comment by rwjblue** [comment]
@Serabe - Thank you for working so hard on this!

**Comment by Serabe** [comment]
Thanks to you, guys, for the support.

**Comment by balinterdi** [comment]
@Serabe Fantastic, you pulled it off! Great work!

**Comment by rwjblue** [comment]
Submitted https://github.com/emberjs/ember.js/pull/12378 with some minor tweaks to the feature flagging syntax.

**Comment by DylanLukes** [comment]
This works great, thanks! Forms with the much-lauded simplified syntax are pretty easy to achieve. 

Utter barebones:

```hbs
{{!-- my-form --}}
<form {{action "submit" on="submit"}}>
  {{yield (hash input=(component 'my-form-field' model=model))}}
</form>
```

```hbs
{{!-- my-form-field --}}
<div>
  {{#if template}}
    {{yield}}  {{!-- escape hatch for custom form fields, they happen --}}
  {{else}}
  <label for="...">{{label}}</label>
    {{input value=(mut (get model for))}}
  {{/if}}
</div>
```

```hbs
{{!-- usage --}}

{{#my-form model=someModel as |form|}}
  {{form.input label="First Name" for="firstName"}}
  {{form.input label="Last Name" for="lastName"}}
  {{#form.input}}
    {{!-- random stuff here --}}
  {{/form.input}}
{{/model-form}}
```

**Pull request merged**